src/main/java/com/onarandombox/MultiverseCore/listeners/MVPlayerListener.java

```diff
@@ -253,112 +243,140 @@ public void playerPortalCheck(PlayerPortalEvent event) {
-        this.plugin.getServer().getScheduler().scheduleSyncDelayedTask(this.plugin,
-                new Runnable() {
-                    @Override
-                    public void run() {
-                        if (!MVPlayerListener.this.pt.playerCanIgnoreGameModeRestriction(world, player)) {
-                            // Check that the player is in the new world and they haven't been teleported elsewhere or the event cancelled.
-                            if (player.getWorld() == world.getCBWorld()) {
-                                Logging.fine("Handling gamemode for player: %s, Changing to %s", player.getName(), world.getGameMode().toString());
-                                Logging.finest("From World: %s", player.getWorld());
-                                Logging.finest("To World: %s", world);
-                                player.setGameMode(world.getGameMode());
-                                // Check if their flight mode should change
-                                // TODO need a override permission for this
-                                if (player.getAllowFlight() && !world.getAllowFlight() && player.getGameMode() != GameMode.CREATIVE) {
-                                    player.setAllowFlight(false);
-                                    if (player.isFlying()) {
-                                        player.setFlying(false);
-                                    }
-                                } else if (world.getAllowFlight()) {
-                                    if (player.getGameMode() == GameMode.CREATIVE) {
-                                        player.setAllowFlight(true);
-                                    }
-                                }
-                            } else {
-                                Logging.fine("The gamemode/allowfly was NOT changed for player '%s' because he is now in world '%s' instead of world '%s'",
-                                        player.getName(), player.getWorld().getName(), world.getName());
-                            }
-                        } else {
-                            Logging.fine("Player: " + player.getName() + " is IMMUNE to gamemode changes!");
-                        }
-                    }
-                }, 1L);
+        this.plugin.getServer().getScheduler().scheduleSyncDelayedTask(
+                this.plugin,
+                ()-> applyGameModeAndFlight(player, world),
+                1L);
+    }
+
+    /**
+     * Applies the gamemode and flight for the specified {@link Player}.
+     *
+     * @param player The {@link Player}.
+     * @param world The world the player is in.
+     */
+    public void applyGameModeAndFlight(@NotNull Player player, @NotNull MVWorld world) {
+        if (MVPlayerListener.this.actionChecker.canKeepGameMode(player, world).isSuccessful()) {
+            Logging.fine("Player: " + player.getName() + " is IMMUNE to gamemode changes!");
+            return;
+        }
```

Comment:
Could a reason for keeping game mode also be that the world's game mode isn't different from the player's current game mode? If this is meant to imply something else maybe the action check method could be named to reflect that better.

Reply:
based on the previous code, there was no such check of whether the gamemode is the same, but ig I can add it.